### PR TITLE
called syncOriginal() after setting attributes getAll() and find() - …

### PR DESCRIPTION
**Pull request by zoul0813**
added support for timestamps, and fixed "isDirty" logic (isDirty is required for timestamps to work properly).

---
This change is [Reviewable](https://reviewable.io/reviews/baopham/laravel-dynamodb/58)

### CODE REVIEW

**Comment by baopham** [comment]
Nice. This change would need some unit tests to make sure it is working and won't break in the future.

If you need help with the tests, let me know. It'd be nice if we can support timestamps.

**Comment by zoul0813** [comment]
Having trouble getting DynamoDbLocal setup for the test.

I already have DDb Local setup for my laravel project, and use the following to start it ... 

`java -Djava.library.path=..path/DynamoDBLocal_lib -jar ..path/DynamoDBLocal.jar -sharedDb --port 8081 -dbPath ..path/webroot/database/local/dynamo `

I adjusted this to the following for the tests

`java -Djava.library.path=..path/DynamoDBLocal_lib -jar ..path/DynamoDBLocal.jar - -port 3000 -dbPath ..path/tmp/laravel-dynamodb/` (my fork is in the tmp folder)

Doing this, I managed to go from having every test generate errors (and take forever trying to connect to what was not there), and now am getting 6 errors ... they all appear to be the same error though.

```
5) BaoPham\DynamoDb\Tests\DynamoDbModelTest::testUpdateRecord
Error: Call to a member function connection() on null

...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Model.php:1042
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Model.php:1008
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Concerns/HasAttributes.php:777
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Concerns/HasAttributes.php:732
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Concerns/HasAttributes.php:525
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Model.php:1285
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Concerns/HasTimestamps.php:71
...path/tmp/laravel-dynamodb/vendor/illuminate/database/Eloquent/Concerns/HasTimestamps.php:42
...path/tmp/laravel-dynamodb/src/DynamoDbModel.php:127
...path/tmp/laravel-dynamodb/src/DynamoDbModel.php:147
...path/tmp/laravel-dynamodb/tests/DynamoDbModelTest.php:67
```

Do I need to do something other than start DDb Local pointing at the project directory? Looks like the tests take the dynamodb_local_init.db and move it to dynamodb_local_test.db which I assume creates all of the necessary tables, with expected default items in it?

**Comment by baopham** [comment]
> the tests take the dynamodb_local_init.db and move it to dynamodb_local_test.db which I assume creates all of the necessary tables 

Yes

Can you make sure you `cd` into the project root and run the exact command:

```
java -Djava.library.path=./DynamoDBLocal_lib -jar dynamodb_local/DynamoDBLocal.jar --port 3000
```

Then in a new tab, run:

```
./vendor/bin/phpunit
```

**Comment by zoul0813** [comment]
Ahh, looks like I actually did have it all setup correctly.  Following the errors, the timestamp logic does apparently try to connect to the database at some point to retrieve a date/time format in the event that one is not already set on the model.  Should be an easy fix to explicitly set the date/time format on DynamoDbModel - just a matter of picking a default format.

I'm thinking we should use the format generated by JavaScript's Date toJSON method, as described here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toJSON

**Comment by baopham** [comment]
Since Laravel is using Carbon so why not use its `toIso8601String` http://carbon.nesbot.com/docs/#api-commonformats?

**Comment by zoul0813** [comment]
I figured using the JavaScript json format would make the json document more compatible with other application code that may be using/accessing it (lambda triggers, client side code, mobile apps, etc) either directly or indirectly.

**Comment by zoul0813** [comment]
Looks like those two formats are the same actually.  Javascripts date object calls toIsoString behind the scenes.  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString

**Comment by baopham** [comment]
ISO 8601 format it is then 👍

**Comment by zoul0813** [comment]
Adding `protected $dateFormat = 'o';` to DynamoDbModel seems to have solved the problem.

```
> phpunit
PHPUnit 5.7.21 by Sebastian Bergmann and contributors.

........................................................          56 / 56 (100%)

Time: 1.84 seconds, Memory: 32.00MB

OK (56 tests, 122 assertions)
```

**Comment by baopham** [comment]
Ah yeah totally forgot about that!

Still need to have 2 test cases though, one for created_at and one for updated_at.

**Comment by zoul0813** [comment]
Not quite sure why, but all the timestamps are being set to 1970-01-01 ... in my app, they are being set to the correct current time.  I checked the HasTimestamps trait and verified that it's creating a new Carbon instance, with the current time ... but for some reason, after it's being set to $this->testModel->created_at it gets set to 1970-01-01 ... not quite sure where this is happening - any ideas? I'm thinking it may have something to do with the tests themselves?

**Comment by baopham** [comment]
Let me try and see

**Comment by baopham** [comment]
For a more deterministic test, maybe we should try `Carbon::setTestNow`

**Comment by zoul0813** [comment]
I created a new test class, so we can setup as many tests as needed - this also allows the other test classes to set `public $timestamps = false;`, as timestamps are on by default and may effect the other test results.

I did use Carbon::setTestNow()

The problem was using `protected $dateFormat = 'o';`, I wasn't paying attention and got that shorthand format off some Microsoft site ... I switched it over to `DateTime::ISO8601` from the PHP Docs...

File: composer.json [hunk 1]
```diff
@@ -23,6 +23,10 @@
     "require-dev": {
         "phpunit/phpunit": "^5.0@dev"
     },
+    "scripts": {
+        "test": "phpunit",
+        "dynamodb_local": "java -Djava.library.path=./DynamoDBLocal_lib -jar dynamodb_local/DynamoDBLocal.jar --port 3000"
+    },
```
Comment (baopham) [inline composer.json]: Very nice! just like `package.json` 😄

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+or
+
+```composer --timeout=0 run dynamodb_local```
+
+Running Tests
+```./vendor/bin/phpunit```
+
+or
+
+```composer run-script tests```
```
Comment (baopham) [inline CONTRIBUTING.md]: Do you mean `test` here?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,3 +1,19 @@
 # Contribution Guide
 
 Thank you for considering contributing to this library. Please make sure your code follows the PSR-2 coding standard and the PSR-4 autoloading standard before sending a pull request.
+
+## Running Tests
```
Comment (baopham) [inline CONTRIBUTING.md]: Might as well remove the test section in README.

File: tests/DynamoDbTimestampTest.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+class DynamoDbTimestampTest extends ModelTest
+{
+    protected function getTestModel()
+    {
+        return new TimestampModel([]);
+    }
+
+    public function testCreateRecord()
+    {
+        \Carbon\Carbon::setTestNow(\Carbon\Carbon::create(2017, 06, 24, 5, 30, 0));
```
Comment (baopham) [inline tests/DynamoDbTimestampTest.php]: Maybe put a `use Carbon\Carbon` on top?

File: tests/DynamoDbTimestampTest.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            'TableName' => $this->testModel->getTable(),
+            'Item' => $item,
+        ]);
+
+        return $item;
+    }
+
+}
+
+class TimestampModel extends \BaoPham\DynamoDb\DynamoDbModel
```
Comment (zoul0813) [inline tests/DynamoDbTimestampTest.php]: Nitpick is really nit-picky... can I just ignore some of these?

Comment (baopham) [inline tests/DynamoDbTimestampTest.php]: Haha yeah, this one can be ignored.

File: CONTRIBUTING.md (hunk 3 above)
Comment (zoul0813) [inline CONTRIBUTING.md]: Ahh, didn't realize there was a test section there - did you add it recently?

Comment (baopham) [inline CONTRIBUTING.md]: No it's been there for a while :)

File: tests/DynamoDbTimestampTest.php (hunk 4 above)
Comment (zoul0813) [inline tests/DynamoDbTimestampTest.php]: yeah, ... that would make sense :)

File: CONTRIBUTING.md (hunk 2 above)
Comment (zoul0813) [inline CONTRIBUTING.md]: I did ... fixing now

File: CONTRIBUTING.md (hunk 3 above)
Comment (zoul0813) [inline CONTRIBUTING.md]: I updated CONTRIBUTING.md and README.md

File: composer.json (hunk 1 above)
Comment (zoul0813) [inline composer.json]: Thanks :)

**Comment by baopham** [comment]
Many thanks @zoul0813 for pushing this through.

**Comment by zoul0813** [comment]
Oh, don't forget to check off the TODO item at the bottom of README.md ... just saw that.

**Comment by baopham** [comment]
Yep one down!

**Pull request merged**